Don't recycle incomplete command entries

cryptorick opened the pull request:
This fixes issue #8.

Don't store incomplete command entries -- like "thun" (which, say, completed to the PATH command "thunar") -- in command history (which gets recycled into subsequent read completion candidates), but still preserve the current functionality of the saving/recyling of "custom commands" (i.e., PATH commands with arguments) like "thunar -v".

The issue of saving incomplete commands was caused by the inner workings of the completing read function (and its dependencies) which saves the user input into the value of a symbol suppiled by the calling code (in our case, `dmenu--history-list`).  Not finding a clear way to "tell" the completing read function exactly which user inputs were OK (and which were not OK) to push onto the history symbol's value, the fix here simply takes that responsibility away from the completing read function and we now just manually jigger our own history.

cryptorick committed:
fix issue #8

Don't store incomplete command entries -- like "thun" (which, say, completed to the PATH command "thunar") -- in command history (which gets recycled into subsequent read completion candidates), but still preserve the current functionality of the saving/recyling of "custom commands" (i.e., PATH commands with arguments) like "thunar -v".

The issue of saving incomplete commands was caused by the inner workings of the completing read function (and its dependencies) which saves the user input into the value of a symbol suppiled by the calling code (in our case, `dmenu--history-list`).  Not finding a clear way to "tell" the completing read function exactly which user inputs were OK (and which were not OK) to push onto the history symbol's value, the fix here simply takes that responsibility away from the completing read function and we now just manually jigger our own history.

lujun9972 commented on dmenu.el:
```diff
@@ -81,18 +81,22 @@ Must be set before initializing Dmenu."
+    ;; Add history entries when the user input was a PATH command followed by
+    ;; one or more arguments.
+    (if (let ((cmd+args (split-string execute-file)))
+          (and (> (length cmd+args) 1)
```
I think the condition of `(> (length cmd+args) 1)` should be removed. otherwise, the command without arguments would not be cached.

cryptorick replied:
Yes, good question.  So, the idea behind this is that the command without arguments is already cached in the list `dmenu--cache-executable-files`.  If we also cache it in `dmenu--history-list` (as the completing read function was doing for us), we then also have to write code to *remove* it from the list of candidates.  That's what (I believe) the `cl-remove-if` call was doing here:
```
(append dmenu--history-list
        (cl-remove-if (lambda (x)
                        (member x dmenu--history-list))
                      dmenu--cache-executable-files))
```
If instead we do NOT admit the command without arguments into `dmenu--history-list` -- which is what this new code is trying to accomplish -- then the `cl-remove-if` call becomes unnecessary.  (See my comment on the next code chunk.)

lujun9972 commented on dmenu.el:
```diff
@@ -81,18 +81,22 @@ Must be set before initializing Dmenu."
+         (execute-file (funcall completing-read-fn
+                                dmenu-prompt-string
+                                (append dmenu--history-list
+                                        dmenu--cache-executable-files)
```
I don't understand why changes here?

cryptorick replied:
(My comment on the other code chunk applies here.)

Since the new code does NOT cache commands without arguments in `dmenu--history-list` anymore, we do not therefore need to filter them out of `dmenu--history-list` (with `cl-remove-if`) as before:
```
(append dmenu--history-list
        (cl-remove-if (lambda (x)
                        (member x dmenu--history-list))
                      dmenu--cache-executable-files))
```
That code now simply becomes:
```
(append dmenu--history-list
        dmenu--cache-executable-files)
```

lujun9972 reviewed (commented):
Thanks, but there are some changes I don't quite understand

lujun9972 commented:
the item in demu-history-list also will be shown ahead. so I think it's better to cache commands without arguments as well.

cryptorick commented:
Ah, I see!  Actually, I *didn't* see before I switched the completing-read function to something else  :) You see, my emacs setup has `ivy-completing-read` as my completing read function -- yours may be different, say, ido's completing read.  With ivy, it doesn't matter if you put the user history at the top of the candidate list because it will reorder the candidate list!   On the other hand, when I switched to ido's completing read, I could see the user input history as the first entries of the completion candidates (that is, ido did NOT reorder the candidates).  So, I will put back most of the original code.  Thanks!

cryptorick committed:
Put back some original code

The commit prior to this assumed that the completing read function would sort the candidate list (given to it).  However, some completing read functions do NOT do this, and when this is the case for some users, they would like to see their particular user history at the top of the completion candidates displayed to them in the UI when `M-x dmenu` is initially invoked.  Hence, some of the original code that handled this is being restored by this commit.

cryptorick commented:
@lujun9972 I (or you) could squash these two commits into one, if you want. Please let me know.  Thanks!

lujun9972 commented:
Since `pushnew` is an alias for `cl-pushnew` in `cl.el`, and `cl.el` is not encouraged to use anymore. I think it will be better to use `cl-pushnew` instead of it.
And also, I‘d like `The most recently used commands come first`, `pushnew` seems unable to implement it.

cryptorick commented:
@lujun9972 yes, you are correct. I will fix!

lujun9972 commented:
@bandali0 It is wired that you type `thun` and it opens the `thunar` but retains the `thun`.

I tried on my machine. I typed `fir`, opened the `firefox`, and the package retained the `firefox` not `fir`

lujun9972 commented:
@cryptorick Do you have the same problem?

cryptorick committed:
Restore more original code

The `pushnew` call will not satisfy the user requirement of putting the most recent commands at the top of the user input history.  The original code (which removes the most recent user input from the history -- no matter how deep in the history it occurs(!) -- and pushes it on to the head of history) does (or rather, did) satisfy that requirement.  Hence, it goes back in.

lujun9972 commented:
If I type a wrong command, it reports an error message that the command can't found, but the wrong command is still retained.
So I think he code `(member (car cmd+args) dmenu--cache-executable-files)))` is useful, we can keep it @cryptorick

cryptorick commented:
@lujun9972 and @bandali0, with these three commits (sorry for 3 of them -- when you squash them, it is only a one line change), when I type "fox", it opened firefox and the history has "firefox" but not "fox".  Same should hold for "thun" and "thunar", namely that "thun" should not be in the history.  Also,"thunar -v" should be retained in history, as was the prior behavior/functionality.

cryptorick commented:
> If I type a wrong command, it reports an error message that the command can't found, but the wrong command is still retained.
So I think he code `(member (car cmd+args) dmenu--cache-executable-files)))` is useful, we can keep it

Ok @lujun9972, I am seeing this here too.  I will fix.

cryptorick committed:
Disallow commands not found on PATH

lujun9972 commented:
@cryptorick 
Thanks a lot. I will merge it :smile:

cryptorick commented:
OK, now up to 4 commits. :)  Oh well, I did indeed learn a lot about how others use dmenu.el and what are some of the subtle and unwritten user requirements.  Thanks for your help and your patience!
